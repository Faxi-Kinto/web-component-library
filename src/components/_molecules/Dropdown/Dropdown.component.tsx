import React, { useState, useEffect, ReactNode, Fragment } from 'react';
import * as Styled from './Dropdown.styles';
import classNames from 'classnames';

export type DropdownOption = {
  text: string;
  value: string;
  icon?: ReactNode;
};

export type DropdownProps<T = {}> = T & {
  optionList: DropdownOption[];
  label?: string;
  description?: string;
  placeholder?: string;
  value?: DropdownOption | string;
  toggleIcon?: JSX.Element;
  toggleIconClassName?: string;
  onChange?: (value: string) => void;
  className?: string;
  optionClassName?: string;
  labelClassName?: string;
  descriptionClassName?: string;
  errorClassName?: string;
  borderColor?: string;
  dropdownOpenBorderColor?: string;
  placeholderColor?: string;
  optionsBackgroundColor?: string;
  optionBorderTopBottomColor?: string;
  optionHoverColor?: string;
  optionSelectedBorderColor?: string;
  optionSelectedBackgroundColor?: string;
  optionSelectedHoverBackgroundColor?: string;
  iconMode?: boolean;
  errorState?: boolean;
};

const Dropdown = <T,>(props: DropdownProps<T>): JSX.Element => {
  const {
    optionList,
    label,
    description,
    placeholder,
    value,
    toggleIcon,
    toggleIconClassName,
    onChange,
    className,
    optionClassName,
    labelClassName,
    descriptionClassName,
    errorClassName,
    borderColor,
    dropdownOpenBorderColor,
    placeholderColor,
    optionsBackgroundColor,
    optionBorderTopBottomColor,
    optionHoverColor,
    optionSelectedBorderColor,
    optionSelectedBackgroundColor,
    optionSelectedHoverBackgroundColor,
    iconMode = false,
    errorState,
  } = props;
  const { error } = (props as any) || { error: '' };

  const [isOpen, setIsOpen] = useState(false);
  const [upwards, setUpwards] = useState(false);

  const [containerRef, setContainerRef] = useState<HTMLDivElement>();
  const [optionsRef, setOptionsRef] = useState<HTMLDivElement>();

  const [currentSelected, setCurrentSelected] = useState(
    (typeof value === 'string'
      ? {
          text: value,
          value: value,
        }
      : value) || {
      text: '',
      value: '',
    }
  );

  useEffect(() => {
    if (optionsRef && containerRef) {
      if (
        containerRef.getBoundingClientRect()?.bottom! +
          optionsRef?.getBoundingClientRect()?.height >
        window.innerHeight!
      ) {
        setUpwards(true);
      } else {
        setUpwards(false);
      }
    }
  }, [containerRef, optionsRef]);

  const renderOptions = () => {
    if (isOpen) {
      return (
        <div
          className={classNames([
            'dropdown-container__options',
            { 'dropdown-container__options--upwards': upwards },
            {
              'dropdown-container__options--overflow-auto':
                optionList.length > 5,
            },
          ])}
          role="list"
          ref={reference => {
            if (reference) {
              setOptionsRef(reference);
            }
          }}
        >
          {optionList.map((option: DropdownOption, index: number) =>
            renderOption(option, index)
          )}
        </div>
      );
    }
  };

  const renderOption = (option: DropdownOption, index: number) => {
    return (
      <div
        key={index}
        onClick={() => {
          setCurrentSelected({ ...option });
          onChange && onChange(option.value);
        }}
        className={classNames([
          'dropdown-container__options__option',
          {
            'dropdown-container__options__option--selected':
              option.value === currentSelected.value,
          },
          optionClassName,
        ])}
        role="listitem"
      >
        {option.icon}
        <span>{option.text}</span>
      </div>
    );
  };

  const handleOpenDropdown = () => {
    setIsOpen(!isOpen);
  };

  return (
    <div className={className}>
      {label && <label className={labelClassName}>{label}</label>}
      <Styled.Container
        ref={reference => {
          if (reference) {
            setContainerRef(reference);
          }
        }}
        onClick={() => handleOpenDropdown()}
        className={classNames([
          'dropdown-container',
          {
            'dropdown-container--icon-mode': iconMode,
          },
          {
            'dropdown-container--open': isOpen,
            'dropdown-container--close': !isOpen,
          },
          {
            'dropdown-container--error': errorState,
          },
          {
            'dropdown-container--unset-min-width': !isOpen && iconMode,
          },
        ])}
        tabIndex={-1}
        onBlur={() => setIsOpen(false)}
        borderColor={borderColor}
        dropdownOpenBorderColor={dropdownOpenBorderColor}
        placeholderColor={placeholderColor}
        optionsBackgroundColor={optionsBackgroundColor}
        optionBorderTopBottomColor={optionBorderTopBottomColor}
        optionHoverColor={optionHoverColor}
        optionSelectedBorderColor={optionSelectedBorderColor}
        optionSelectedBackgroundColor={optionSelectedBackgroundColor}
        optionSelectedHoverBackgroundColor={optionSelectedHoverBackgroundColor}
      >
        {currentSelected.text === '' ? (
          placeholder ? (
            <div className="dropdown-container__placeholder">{placeholder}</div>
          ) : (
            <Fragment>
              {optionList[0].icon}
              {!iconMode && <div>{optionList[0].text}</div>}
            </Fragment>
          )
        ) : (
          <Fragment>
            {currentSelected.icon}
            {!iconMode && <div>{currentSelected.text}</div>}
          </Fragment>
        )}

        {toggleIcon &&
          React.cloneElement(toggleIcon, {
            className: classNames([
              'dropdown-container__arrow',
              toggleIconClassName,
            ]),
          })}
        {renderOptions()}
      </Styled.Container>
      {description && <div className={descriptionClassName}>{description}</div>}
      {errorState && <div className={errorClassName}>{error}</div>}
    </div>
  );
};

export default Dropdown;
