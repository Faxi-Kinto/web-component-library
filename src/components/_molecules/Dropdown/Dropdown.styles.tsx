import styled from 'styled-components';
import {
  flexRow,
  pxToRem,
  positionAbsolute,
  flexColumn,
} from '@faxi/web-css-utilities';

export type DropdownStylingProps = {
  borderColor?: string;
  backgroundColor?: string;
  dropdownOpenBorderColor?: string;
  dropdownErrorBorderColor?: string;
  placeholderColor?: string;
  optionsBackgroundColor?: string;
  optionBorderTopBottomColor?: string;
  optionHoverColor?: string;
  optionSelectedBorderColor?: string;
  optionSelectedBackgroundColor?: string;
  optionSelectedHoverBackgroundColor?: string;
};

export const Container = styled.div<DropdownStylingProps>`
  ${flexRow('space-between', 'center')};
  height: ${pxToRem('45px')};
  position: relative;
  min-width: ${pxToRem('250px')};
  border: ${pxToRem('1px')} solid
    ${(props: DropdownStylingProps) =>
      props.borderColor ? props.borderColor : 'black'};
  background: ${(props: DropdownStylingProps) =>
    props.backgroundColor ? props.backgroundColor : 'white'};
  padding: ${pxToRem('10px')} ${pxToRem('30px')};
  cursor: pointer;
  font-size: ${pxToRem('18px')};

  &:focus {
    outline: 0;
  }

  .dropdown-container {
    &__placeholder {
      color: ${(props: DropdownStylingProps) =>
        props.placeholderColor ? props.placeholderColor : 'black'};
      padding-right: ${pxToRem('12px')};
    }

    &__selected-main {
      padding-right: ${pxToRem('12px')};
    }

    &__options {
      z-index: 1;
      position: absolute;
      top: ${pxToRem('44px')};
      right: 0;
      border: ${pxToRem('1px')} solid
        ${(props: DropdownStylingProps) =>
          props.borderColor ? props.borderColor : 'black'};
      border-top: none;
      background: ${(props: DropdownStylingProps) =>
        props.optionsBackgroundColor ? props.optionsBackgroundColor : 'white'};
      width: 100%;
      ${flexColumn()};

      &--upwards {
        ${positionAbsolute('inherit', '0', pxToRem('44px'), '0')};
        border-top: ${pxToRem('1px')} solid
          ${(props: DropdownStylingProps) =>
            props.borderColor ? props.borderColor : 'black'};
        border-bottom: none;
      }

      &--overflow-auto {
        height: ${pxToRem('282px')};
        overflow-y: auto;
        overflow-x: hidden;
      }

      &__option {
        border-bottom: ${pxToRem('1px')} solid
          ${(props: DropdownStylingProps) =>
            props.optionBorderTopBottomColor
              ? props.optionBorderTopBottomColor
              : 'gray'};

        &:last-of-type {
          border-bottom: none;
        }

        cursor: pointer;
        padding: ${pxToRem('10px')} ${pxToRem('30px')};
        ${flexRow('space-between', 'center')};
        word-break: break-all;

        &:hover {
          background: ${(props: DropdownStylingProps) =>
            props.optionHoverColor ? props.optionHoverColor : 'gray'};
        }

        &--selected {
          border: ${pxToRem('1px')} solid
            ${(props: DropdownStylingProps) =>
              props.optionSelectedBorderColor
                ? props.optionSelectedBorderColor
                : 'black'};
          margin: ${pxToRem('-1px')} ${pxToRem('-1px')} 0;
          background: ${(props: DropdownStylingProps) =>
            props.optionSelectedBackgroundColor
              ? props.optionSelectedBackgroundColor
              : 'lightgray'};

          &:hover {
            background: ${(props: DropdownStylingProps) =>
              props.optionSelectedHoverBackgroundColor
                ? props.optionSelectedHoverBackgroundColor
                : 'gray'};
          }
        }
      }
    }
  }

  &.dropdown-container {
    &--open {
      border: ${pxToRem('1px')} solid
        ${(props: DropdownStylingProps) =>
          props.dropdownOpenBorderColor
            ? props.dropdownOpenBorderColor
            : 'gray'};

      .dropdown-container__arrow {
        transform: rotate(180deg);
      }
    }

    &--error {
      border: ${pxToRem('1px')} solid
        ${(props: DropdownStylingProps) =>
          props.dropdownErrorBorderColor
            ? props.dropdownErrorBorderColor
            : 'gray'};
    }

    &--icon-mode {
      border: none;
      width: fit-content;

      .dropdown-container__options {
        > div:first-child {
          border-top: ${pxToRem('1px')} solid
            ${(props: DropdownStylingProps) =>
              props.optionBorderTopBottomColor
                ? props.optionBorderTopBottomColor
                : 'gray'};
        }
      }
    }

    &--unset-min-width {
      min-width: unset;
    }
  }
`;
