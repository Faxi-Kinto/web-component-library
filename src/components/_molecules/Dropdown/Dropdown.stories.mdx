import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import Dropdown from './';

<Meta title="Core|Molecules/Dropdown" />

# Dropdown Component

## Implementation

#### Default

<Preview>
  <Story name="default">
    <div style={{display: 'flex'}}>
    <Dropdown
      type="select"
      options={[
        {
          label: 'Option 1',
          value: 'o1',
        },
        { label: 'option2option2option2option2option2', value: 'o2' },
        { label: 'option3', value: 'op3' },
      ]}
      renderInBody={false}
    />
    </div>
  </Story>
</Preview>

#### With placeholder

<Preview>
  <Story name="placeholder">
    <Dropdown
      type="select"
      options={[
        { label: 'Show all members', value: 'o1' },
        { label: 'option2option2option2option2option2', value: 'o2' },
        { label: 'option3', value: 'op3' },
      ]}
      placeholder="Placeholder"
    />
  </Story>
</Preview>

#### Icon mode, with label

<Preview>
  <Story name="expander">
    <Dropdown
      type="expander"
      options={[
        {
          label: 'Option 1',
          value: 'o1',
        },
        {
          label: 'Option 2',
          value: 'o2',
        },
        {
          label: 'Option 3',
          value: 'op3',
        },
      ]}
      label="Icon mode here"
      iconMode
    />
  </Story>
</Preview>

## Props

<Props of={Dropdown} />
