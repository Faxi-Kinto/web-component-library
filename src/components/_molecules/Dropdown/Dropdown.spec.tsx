import React from 'react';
import renderer from 'react-test-renderer';
import Dropdown from './Dropdown.component';

test('Dropdown renders correctly', () => {
  const component = renderer.create(
    <Dropdown
      options={[
        { label: 'Show all members', value: 'o1' },
        { label: 'option2option2option2option2option2', value: 'o2' },
        { label: 'option3', value: 'op3' },
      ]}
      placeholder="Placeholder"
      type="select"
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
