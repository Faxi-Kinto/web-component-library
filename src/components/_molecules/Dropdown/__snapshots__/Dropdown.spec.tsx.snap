// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 2.8125rem;
  position: relative;
  min-width: 15.625rem;
  border: 0.0625rem solid black;
  padding: 0.625rem 1.875rem;
  cursor: pointer;
  font-size: 1.125rem;
}

.c0:focus {
  outline: 0;
}

.c0 .dropdown-container__placeholder {
  color: black;
}

.c0 .dropdown-container__arrow {
  margin-left: 0.9375rem;
}

.c0 .dropdown-container__options {
  position: absolute;
  top: 2.75rem;
  right: 0;
  border: 0.0625rem solid black;
  border-top: none;
  background: white;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: initial;
  -webkit-justify-content: initial;
  -ms-flex-pack: initial;
  justify-content: initial;
  -webkit-align-items: initial;
  -webkit-box-align: initial;
  -ms-flex-align: initial;
  align-items: initial;
}

.c0 .dropdown-container__options--upwards {
  position: absolute;
  top: inherit;
  right: 0;
  bottom: 2.75rem;
  left: 0;
  border-top: black;
  border-bottom: none;
}

.c0 .dropdown-container__options--overflow-auto {
  height: 17.625rem;
  overflow-y: auto;
  overflow-x: hidden;
}

.c0 .dropdown-container__options__option {
  border-bottom: 0.0625rem solid gray;
  cursor: pointer;
  padding: 0.625rem 1.875rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  word-break: break-all;
}

.c0 .dropdown-container__options__option:last-of-type {
  border-bottom: none;
}

.c0 .dropdown-container__options__option:hover {
  background: gray;
}

.c0 .dropdown-container__options__option--selected {
  border: 0.0625rem solid black;
  margin: -0.0625rem -0.0625rem 0;
  background: lightgray;
}

.c0 .dropdown-container__options__option--selected:hover {
  background: gray;
}

.c0.dropdown-container--open {
  border: 0.0625rem solid gray;
}

.c0.dropdown-container--open .dropdown-container__arrow {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c0.dropdown-container--icon-mode {
  border: none;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c0.dropdown-container--icon-mode .dropdown-container__options > div:first-child {
  border-top: 0.0625rem solid gray;
}

.c0.dropdown-container--unset-min-width {
  min-width: unset;
}

<div
  className="c0 dropdown-container dropdown-container--close"
  onBlur={[Function]}
  onClick={[Function]}
  tabIndex={-1}
>
  <div
    className="dropdown-container__placeholder"
  >
    Placeholder
  </div>
</div>
`;
